bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java

```diff
@@ -635,4 +640,16 @@ protected void changeThingType(ThingTypeUID thingTypeUID, Configuration configur
         }
     }
 
+    public void setChannelsFromThingTypeDefinition(ThingType thingType,
+            ConfigDescriptionRegistry configDescriptionRegistry) {
+        logger.info("Synchronizing channels for thing {}", getThing().getUID());
+        List<Channel> channels = ThingFactoryHelper.createChannels(thingType, getThing().getUID(),
+                configDescriptionRegistry);
+
+        ThingBuilder thingBuilder = editThing();
+        thingBuilder.withChannels(channels);
```

Comment:
IIRC this method clears the list of all existing channels. We have to filter the list of new channels to add only non existing ones. On the other hand we have to deal with removal of channels and changes in channel types too.

Be aware there currently is an issue regarding the `withoutChannels` for the `BridgeBuilder` - it returns a `ThingBuilder` instance - see https://github.com/openhab/openhab-core/pull/908/files#diff-44fdf6ebca4daa065ec6088297fb2859R82.

Reply:
> We have to filter the list of new channels to add only non existing ones

Why ?

Indeed this method clears all channels on the thing and replaces them with the one declared in the XML, but once the handler is initialized again - all other channels (added dynamically) will be re-added (it will behave just like if the thing is paired for a very first time)

Reply:
Yes, exactly but please keep in mind that channels can have a user specific configuration. You will throw away any of them when you re-add all the channels. We should try to keep them.

Reply:
and there are channel that can be added by user interaction -- so manually per request. That information will be lost, too.

Reply:
Yes - I know - this is already mentioned in the description in the PR.

One question on that topic:
How are actually those channels used ?
How are the stated updated if the handler doesn't know about them. 
How commands sent to those channels handled ?